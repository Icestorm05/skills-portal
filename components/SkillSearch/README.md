# Skill Search

## Functionality

The skill search component displays a search input field, which allows for multiple queries to be entered.

## Usage

### Props

`<skill-search>` supports the following custom component attributes:

| attribute | type | description
| --- | --- | ---
| `autocomplete-items` | Array *optional* | The array of autocomplete items. Can either contain an array of strings, or an array of objects.
| `autocomplete-item-key` | String *optional* | The key containing the value to display for each object in the `autocomplete-items` array. The value should be a string.
| `autocomplete-item-tags-key` | String *optional* | The key containing the tags to display for each item. The value should be an array of strings.
| `search-value` | Array | The default value of the input field. Can also be bound to using `v-model`.
| `legend` | Object | The legend to apply colors to. Is used to map autocomplete tags to colors.

### Events

`<skill-search>` emits the following events:

| event | type | description
| --- | --- | ---
| `search-value-updated` | Array | The updated search values. If the `search-value` prop was bound to using `v-model`, the variable will automatically update.

### Example

```html
<skill-search v-model="filters"
              :autocomplete-items="autocompleteItems"
              autocomplete-item-key="name"
              autocomplete-item-tags-key="tags"
              :legend="legend"
              @search-value-updated="filtersUpdated" />
```

```ts
import Component from 'nuxt-class-component';
import Vue from 'vue';

import SkillSearch from '~/components/SkillSearch/SkillSearch.vue';

@Component({
    components: {
        SkillSearch,
    },
})
export default class GenericComponent extends Vue {
    private autocompleteItems = [{
        name: 'Item 1',
        tags: ['Parent 1', 'Parent 2'],
    }, {
        name: 'Item 2',
        tags: ['Parent 1', 'Parent 2'],
    }];
    private legend = {
        'Parent 1': 'blue',
        'Parent 2': 'red',
    };
    private filters = [];

    private filtersUpdated(updatedFilters) {
        updatedFilters.forEach(updatedFilter => {
            console.log(updatedFilter.name);
            console.log(updatedFilter.tags);
        });
    }
}
```
