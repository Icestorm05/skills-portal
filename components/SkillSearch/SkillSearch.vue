<template lang="pug">
    #SkillSearchContainer
        v-select(prepend-icon="search"
                label="Skill Search"
                :autocomplete="autocompleteItems.length ? true : false"
                browser-autocomplete="off"
                clearable
                autofocus
                solo
                chips
                tags
                multiple
                hide-selected
                :item-text="autocompleteItemKey"
                :item-value="autocompleteItemKey"
                :return-object="autocompleteItemKey ? true : false"
                ref="select"
                :items="autocompleteItems"
                :value="searchValue"
                :filter="filter"
                @input="searchValueChanged").mb-3#SkillSearch.SkillSearch
            template(slot="item" slot-scope="{ item }")
              v-layout(align-center
                       row
                       wrap)
                v-flex
                  span(v-if="item[autocompleteItemKey]") {{ item[autocompleteItemKey] }}
                  span(v-else) {{ item }}
                skill-search-tag(v-if="Array.isArray(item[autocompleteItemTagsKey])"
                                 v-for="parent in item[autocompleteItemTagsKey]"
                                 :key="parent"
                                 :legend="legend"
                                 :text="parent"
                                 :tags="item[autocompleteItemTagsKey]").hidden-xs-only
            template(slot="selection" slot-scope="{ item }")
              skill-search-tag(close
                               :legend="legend"
                               :text="item[autocompleteItemKey] ? item[autocompleteItemKey] : item"
                               :tags="item[autocompleteItemTagsKey]"
                               prioritize-parent-color
                               :autocomplete-items="autocompleteItems"
                               :autocomplete-item-tags-key="autocompleteItemTagsKey"
                               @close-tag="remove(item)")
                template(v-if="Array.isArray(item[autocompleteItemTagsKey])"
                         v-for="parent in item[autocompleteItemTagsKey]")  ({{ parent }})
</template>

<script lang="ts">
import Component from "nuxt-class-component";
import Vue from "vue";
import { Model, Prop, Watch } from "vue-property-decorator";

import { ISkillCategoryLegend } from "types/legend";
import { ISkillQuery } from "types/skillQuery";

import theme from "~/assets/json/theme.json";

import { VFlex, VLayout } from "vuetify/es5/components/VGrid";
import VSelect from "vuetify/es5/components/VSelect";
import SkillSearchTag from "./SkillSearchTag/SkillSearchTag.vue";

interface IAny {
  [key: string]: string | string[];
}

@Component({
  components: {
    SkillSearchTag,
    VFlex,
    VLayout,
    VSelect
  }
})
export default class SkillSearch extends Vue {
  @Prop({ type: Array, default: () => [] })
  private autocompleteItems: Array<IAny | string>;
  @Prop({ type: String })
  private autocompleteItemKey: string | undefined;
  @Prop({ type: String })
  private autocompleteItemTagsKey: string | undefined;
  @Model("search-value-changed")
  @Prop({ type: Array, default: () => [] })
  private searchValue: Array<IAny | string>;
  @Prop({ type: Object })
  private legend: ISkillCategoryLegend | undefined;

  private filter(
    item: IAny | string,
    queryText: string | undefined,
    itemText: string | {}
  ) {
    if (queryText) {
      const itemTextStr =
        typeof itemText !== "string" ? JSON.stringify(itemText) : itemText;
      const match = itemTextStr.match(new RegExp(this.escape(queryText), "ig"))
        ? true
        : false;

      const path =
        typeof item !== "string" &&
        (this.autocompleteItemTagsKey ? true : false)
          ? item[this.autocompleteItemTagsKey as string]
          : undefined;
      const pathMatch =
        path && Array.isArray(path)
          ? path.some(
              child =>
                child.match(new RegExp(this.escape(queryText), "ig"))
                  ? true
                  : false
            )
          : false;

      return match || pathMatch;
    } else {
      return true;
    }
  }

  private escape(text: string) {
    return text.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, "\\$&");
  }

  private searchValueChanged(updatedQueries: Array<string | IAny>) {
    (this.$refs.select as VSelect).hideMenu();
    this.$emit("search-value-changed", updatedQueries);
  }

  private remove(valueToRemove: ISkillQuery | string) {
    (this.$refs.select as VSelect).hideMenu();
    const filteredSearchValue = this.searchValue.filter(value => {
      const v = typeof value === "string" ? value : value.skill;
      const vToRemove =
        typeof valueToRemove === "string" ? valueToRemove : valueToRemove.skill;
      return v !== vToRemove;
    });
    this.$emit("search-value-changed", filteredSearchValue);
  }

  @Watch("autocompleteItems")
  private autocompleteItemsChanged() {
    this.$emit(
      "search-value-changed",
      this.searchValue.filter(value => {
        const query =
          typeof value === "object"
            ? this.autocompleteItemKey
              ? value[this.autocompleteItemKey]
              : undefined
            : value;

        return this.autocompleteItems.some(autocompleteItem => {
          if (typeof autocompleteItem === "object") {
            const autocompleteItemQuery = this.autocompleteItemKey
              ? autocompleteItem[this.autocompleteItemKey]
              : undefined;
            const autocompleteItemTags = this.autocompleteItemTagsKey
              ? autocompleteItem[this.autocompleteItemTagsKey]
              : undefined;

            if (
              typeof query === "string" &&
              typeof autocompleteItemQuery === "string"
            ) {
              const queryMatch =
                typeof value === "string"
                  ? autocompleteItemQuery.match(
                      new RegExp(this.escape(query), "ig")
                    )
                    ? true
                    : false
                  : query === autocompleteItemQuery;
              const tagMatch =
                Array.isArray(autocompleteItemTags) &&
                autocompleteItemTags.some(tag => {
                  return tag.match(new RegExp(this.escape(query), "ig"))
                    ? true
                    : false;
                });
              return queryMatch || tagMatch;
            } else {
              return false;
            }
          } else {
            return typeof query === "string"
              ? autocompleteItem.match(new RegExp(this.escape(query), "ig"))
                ? true
                : false
              : false;
          }
        });
      })
    );
  }
}
</script>

<style lang="stylus" scoped>
#SkillSearchContainer {
  display: flex;
  flex-direction: column;
}

.SkillSearch {
  flex-basis: auto;

  /deep/ .input-group__input {
    flex: 1;
  }
}

.SkillParent {
  font-size: 0.8em;
}
</style>
