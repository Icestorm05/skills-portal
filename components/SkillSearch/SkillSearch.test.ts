import SkillSearch from "./SkillSearch.vue";
import { mount } from "@vue/test-utils";

// Add a wrapping <div data-app="true"></div> to silence Vuetify warnings.
const app = document.createElement("div");
app.setAttribute("data-app", "true");
document.body.appendChild(app);

declare const expect: jest.Expect;

describe("Skill Search", () => {
  test("is a Vue instance", () => {
    const wrapper = mount(SkillSearch);
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test("snapshot has not changed", async () => {
    const wrapper = mount(SkillSearch, {
      propsData: { searchValue: ["test123"] }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("emits nothing by default", () => {
    const wrapper = mount(SkillSearch);
    expect(wrapper.emitted("search-value-changed")).toBeFalsy();
  });
});
