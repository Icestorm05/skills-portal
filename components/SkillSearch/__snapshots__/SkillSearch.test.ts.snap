// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skill Search snapshot has not changed 1`] = `
<div id="SkillSearchContainer">
    <div tabindex="-1" data-uid="8" class="mb-3 SkillSearch input-group input-group--append-icon input-group--prepend-icon input-group--solo input-group--text-field input-group--select input-group--autocomplete input-group--single-line input-group--chips input-group--multiple primary--text">
        <label for="SkillSearch">Skill Search</label>
        <div class="input-group__input">
            <i aria-hidden="true" class="icon material-icons input-group__prepend-icon">search</i>
            <div class="input-group__selections" style="overflow: hidden;">
                <input autofocus="autofocus" id="SkillSearch" tabindex="0" role="combobox" class="input-group--select__autocomplete" autocomplete="off">
            </div>
            <div class="menu" style="display: inline-block;"></div>
            <i aria-hidden="true" class="icon material-icons input-group__append-icon input-group__icon-cb">arrow_drop_down</i>
        </div>
        <div class="input-group__details">
            <!---->
        </div>
    </div>
</div>
`;
